test/failure_app_test.rb

```diff
@@ -232,6 +232,33 @@ def call_failure(env_params={})
         assert_equal 'http://test.host/users/sign_in.js', @response.second["Location"]
       end
     end
+
+    test 'returns to configured authentication_failure_route with symbol' do
+      swap Devise, authentication_failure_route: :root_path do
+        call_failure
+        assert_equal 302, @response.first
+        assert_equal 'You need to sign in or sign up before continuing.', @request.flash[:alert]
+        assert_equal 'http://test.host/', @response.second['Location']
+      end
+    end
+
+    test 'returns to root if configured authentication_failure_route not exists' do
```

Comment:
Isn't it better to raise a 404? Users can have a chance to fix it and we don't hide unexpected behavior. What do you think?